domain/status/state/storage.go

```diff
@@ -0,0 +1,335 @@
+		// Get current status.
+		currentStatus, isProvisioned, err := st.getFilesystemProvisioningStatus(ctx, tx, filesystemUUID)
+		if err != nil && !errors.Is(err, statusNotFound) {
+			return errors.Errorf("getting current filesystem status: %w", err)
+		}
+		if err == nil {
+			// Check we can transition from current status to the new status.
+			err = status.FilesystemStatusTransitionValid(currentStatus, isProvisioned, sts)
+			if err != nil {
+				return errors.Capture(err)
+			}
+		}
+
+		return st.updateFilesystemStatus(ctx, tx, filesystemUUID, sts)
+	})
+	if err != nil {
+		return errors.Errorf("updating filesystem status for %q: %w", filesystemUUID, err)
+	}
+	return nil
+}
+
+// getFilesystemStatus gets the status of the given filesystem
+// and a bool indicating if it is provisioned.
+// The following errors can be expected:
+// - [storageerrors.FilesystemNotFound] if the filesystem doesn't exist.
+func (st *State) getFilesystemProvisioningStatus(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid storage.FilesystemUUID,
+) (status.StorageFilesystemStatusType, bool, error) {
+	id := filesystemUUID{
+		FilesystemUUID: uuid.String(),
+	}
+	var sts storageProvisioningStatusInfo
+
+	stmt, err := st.Prepare(`
+SELECT    &storageProvisioningStatusInfo.*
+FROM      storage_filesystem sf
+LEFT JOIN storage_filesystem_status sfs ON sf.uuid = sfs.filesystem_uuid
+LEFT JOIN storage_instance_filesystem sif ON sf.uuid = sif.storage_filesystem_uuid
+WHERE     sf.uuid = $filesystemUUID.uuid
+`, id, sts)
+	if err != nil {
+		return -1, false, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, id).Get(&sts)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return -1, false, storageerrors.FilesystemNotFound
+	} else if err != nil {
+		return -1, false, errors.Capture(err)
+	}
+	if !sts.StatusID.Valid {
+		return -1, false, statusNotFound
+	}
+	statusType, err := status.DecodeStorageFilesystemStatus(int(sts.StatusID.Int16))
+	if err != nil {
+		return -1, false, errors.Capture(err)
+	}
+	isProvisioned := sts.StorageInstanceUUID != ""
+	return statusType, isProvisioned, nil
+}
+
+// ImportFilesystemStatus sets the given filesystem status.
+// The following errors can be expected:
+// - [storageerrors.FilesystemNotFound] if the filesystem doesn't exist.
+func (st *State) ImportFilesystemStatus(
+	ctx context.Context,
+	filesystemID string,
+	sts status.StatusInfo[status.StorageFilesystemStatusType],
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		filesystemUUID, err := st.getFilesystemUUIDByID(ctx, tx, filesystemID)
+		if err != nil {
+			return errors.Errorf("getting filesystem UUID: %w", err)
+		}
+		return st.updateFilesystemStatus(ctx, tx, filesystemUUID, sts)
+	})
+}
+
+func (st *State) getFilesystemUUIDByID(
+	ctx context.Context,
+	tx *sqlair.TX,
+	id string,
+) (storage.FilesystemUUID, error) {
+	arg := filesystemID{ID: id}
```

Comment:
The convention for naming this based on the other parts of the domain would be `filesystemUUIDID`.